### PR TITLE
Refactor JavaScript for new Credit Card workflow [TPWDCEE-4890]

### PR DESCRIPTION
**Pull request by pretzelhands**
I suggest reviewing this pull request in your IDE as the JavaScript has been rewritten from scratch

### CODE REVIEW

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 1]
```diff
@@ -35,6 +36,7 @@ public function initContent()
     {
         $this->ajax = true;
         $this->vaultModel = new CreditCardVault($this->context->customer->id);
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Please rename it to CreditCardVault

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 2]
```diff
@@ -45,6 +47,11 @@ public function initContent()
      */
     public function displayAjaxListStoredCards()
     {
+        $templatePath = join(
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: join is an alias method please use implode

Comment (tomazpu) [reply wirecardpaymentgateway/controllers/front/creditcard.php]: I would also consider a helper for returning template paths. The same is happening for FormPost, PIA, others.

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 3]
```diff
@@ -45,6 +47,11 @@ public function initContent()
...
+            DIRECTORY_SEPARATOR,
+            [_PS_MODULE_DIR_, \WirecardPaymentGateway::NAME, 'views', 'templates', 'front', 'creditcard_list.tpl']
+        );
+
         $data = [
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: I think to create a private method for creating data is better, as this method is just fetching the template with data.

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 4]
```diff
@@ -62,18 +72,22 @@ public function displayAjaxListStoredCards()
      *
      * @since 1.1.0
      */
-    public function displayAjaxAddCard()
+    public function displayAjaxSaveCard()
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: The method description is wrong to what it is doing.

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 5]
```diff
@@ -41,7 +41,7 @@ public function postProcess()
         $this->context->cookie->__set('pia-enabled', false);
         $shopConfigService = new ShopConfigurationService($paymentType);
         $cartId = \Tools::getValue('order_number');
-        $cart = new Cart($cartId);
+        $cart = $cartId !== false ? new Cart($cartId) : $this->context->cart;
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/front/payment.php]: please break out of the namespace in front of the cart even tho at the time we have no namespace but just to be consistent.

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 6]
```diff
@@ -7,289 +7,413 @@
...
+var Constants = {
+    IFRAME_HEIGHT_DESKTOP: 410,
+    IFRAME_HEIGHT_MOBILE: 390,
+    IFRAME_HEIGHT_CUTOFF: 992,
+
+    MODAL_ID: "#wirecard-ccvault-modal",
+    IFRAME_ID: "#wirecard-integrated-payment-page-frame",
+    CONTAINER_ID: "payment-processing-gateway-credit-card-form",
+    PAYMENT_FORM_ID: "form[action*=\"creditcard\"]",
+    PAYMENT_METHOD_RADIO_ID: "input[name=\"payment-option\"][data-module-name=\"wd-creditcard\"]",
```
Comment (tomazpu) [inline wirecardpaymentgateway/views/js/creditcard.js]: Please rename the variable to something more specific to credit card

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 7]
```diff
@@ -7,289 +7,413 @@
...
 
-        function logCallback(response)
-        {
-            jQuery(document).off("submit", "#payment-form");
+/**
+ * Initializes all event handlers for the interface
+ *
+ * @since 2.4.0
+ */
+function initializeHandlers()
```
Comment (tomazpu) [inline wirecardpaymentgateway/views/js/creditcard.js]: Rename to more specific what do you initialize that it's for CC

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 8]
```diff
@@ -7,289 +7,413 @@
...
+ *
+ * @param error
+ * @since 2.4.0
+ */
+function onFormError(error)
+{
+    console.error("Form error:", error);
 
+    initializeForm();
+}
```
Comment (tomazpu) [inline wirecardpaymentgateway/views/js/creditcard.js]: Missing newline at the end of file

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 9]
```diff
@@ -7,289 +7,413 @@
...
+ * @since 2.4.0
+ */
+function onFormDataReceived(formData)
+{
+    var $form = jQuery(Constants.PAYMENT_FORM_ID);
+
+    // WPP requires the parameter in snake_case
+    // eslint-disable-next-line camelcase
+    formData.wpp_options_cvv_hidden = true;
+    attachFormField($form, "order_number", formData.field_value_1);
```
Comment (tomazpu) [inline wirecardpaymentgateway/views/js/creditcard.js]: field_value_1 is the custom field cart_id why is it named order_number?

File: wirecardpaymentgateway/controllers/front/creditcard.php (hunk 2 above)
Comment (pretzelhands) [reply wirecardpaymentgateway/controllers/front/creditcard.php]: New helper has been extracted for `$templatePath`. `join` has been changed to `implode`

File: composer.json [hunk 10]
```diff
@@ -7,7 +7,7 @@
   "require": {
     "php": ">=5.6",
     "symfony/options-resolver": "^3",
-    "wirecard/payment-sdk-php": "3.8.0",
+    "wirecard/payment-sdk-php": "dev-PRBUG-2115-examples",
```
Comment (tomazpu) [inline composer.json]: This will be updated

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 11]
```diff
@@ -45,52 +49,93 @@ public function initContent()
...
-        $data = [
-            'cards' => $this->vaultModel->getUserCards($this->context->cart->id_address_invoice),
-            'strings' => [
-                'use' => $this->getTranslatedString('vault_use_card_text'),
-                'delete' => $this->getTranslatedString('vault_delete_card_text')
-            ]
-        ];
+        $templatePath = TemplateHelper::getTemplatePath('creditcard_list');
+
+        $this->context->smarty->assign($this->getCardListData());
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: I would put the getCardListData to the model, to prepare the data for the controller.

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 12]
```diff
@@ -45,52 +49,93 @@ public function initContent()
...
-            $this->displayAjaxListStoredCards();
+        if (!$tokenId || !$maskedPan) {
+            $response = new Response("No token or PAN provided.", 400);
+            $response->send();
+
+            return;
         }
 
-        $this->vaultModel->addCard($maskedpan, $tokenId, $this->context->cart->id_address_invoice);
+        $this->credit_card_vault->addCard($maskedPan, $tokenId, $this->context->cart->id_address_invoice);
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: can we rename credit_card_valut to credit_card_valut_model? I think its better for reading the code. We save no expiration date?

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 13]
```diff
@@ -67,8 +67,8 @@ public function postProcess()
      */
     private function determineFinalOrderId()
     {
-        // $cartId used for order_number within intial request
-        $cartId = \Tools::getValue('order_number');
+        // $cartId used for cart_id within intial request
+        $cartId = \Tools::getValue('cart_id');
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/front/payment.php]: here is no need for an else in line 77, or am i wrong?

File: wirecardpaymentgateway/helper/TemplateHelper.php [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+     *
+     * @param string $template
+     * @return string
+     * @since 2.4.0
+     */
+    public static function getTemplatePath($template)
+    {
+        return implode(
+            DIRECTORY_SEPARATOR,
+            [_PS_MODULE_DIR_, \WirecardPaymentGateway::NAME, 'views', 'templates', 'front', $template . '.tpl']
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/TemplateHelper.php]: The path is now hardcoded to frontend templates, but nowhere in the method it is stated, please also create constants for the strings.

File: wirecardpaymentgateway/helper/TransactionBuilder.php [hunk 15]
```diff
@@ -185,8 +185,8 @@ private function addCustomField($key, $value)
      */
     private function addTokenId()
     {
-        if ($this->transaction instanceof CreditCardTransaction && \Tools::getValue('token_id')) {
-            $this->transaction->setTokenId(\Tools::getValue('token_id'));
+        if ($this->transaction instanceof CreditCardTransaction && \Tools::getValue('tokenId')) {
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/TransactionBuilder.php]: all params are in snake case here i see camel case can we change that?

**Comment by tomazpu** [comment]
I found an issue. If you search for the class OrderManager you will see the constructor expects no params, but the use in classes TransactionBuilder and wirecardpaymentgateway.php happens with one parameter. Please fix it.

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/controllers/front/payment.php  1
- wirecardpaymentgateway/helper/TemplateHelper.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/views/js/creditcard.js  -1
         

Clones removed
==============
+ wirecardpaymentgateway/views/js/creditcard.js  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4399691&bid=14911644)

File: wirecardpaymentgateway/controllers/front/creditcard.php (hunk 11 above)
Comment (pretzelhands) [reply wirecardpaymentgateway/controllers/front/creditcard.php]: Did some refactoring which makes this method obsolete entirely

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 5
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/controllers/front/payment.php  1
- wirecardpaymentgateway/helper/TemplateHelper.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/views/js/creditcard.js  -1
         

Clones removed
==============
+ wirecardpaymentgateway/views/js/creditcard.js  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4399691&bid=14911644)

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 16]
```diff
@@ -7,289 +7,410 @@
...
+ * @since 2.4.0
+ */
+function saveCardAndSubmitToShop(tokenId, maskedPan)
+{
+    var cardSavingRequest = jQuery.ajax({
+        url: ccControllerUrl,
+        data: {
+            action: "saveCard",
+            token_id: tokenId,
+            masked_pan: maskedPan
```
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Identifier 'masked_pan' is not in camel case. (camelcase)](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4399691)

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 17]
```diff
@@ -7,289 +7,410 @@
...
+ */
+function getFormData(tokenId = null)
+{
+    var formDataRequest = jQuery.ajax({
+        url: ccControllerUrl,
+        dataType: "json",
+        data: {
+            action: "getSeamlessConfig",
+            cart_id: cartId,
+            token_id: tokenId
```
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Identifier 'token_id' is not in camel case. (camelcase)](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4399691)

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 18]
```diff
@@ -7,289 +7,410 @@
...
+ * @param cardId
+ * @since 2.4.0
+ */
+function deleteCard(cardId)
+{
+    var cardDeletionRequest =  jQuery.ajax({
+        url: ccControllerUrl,
+        data: {
+            action: "deleteCard",
+            card_id: cardId
```
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Identifier 'card_id' is not in camel case. (camelcase)](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4399691)

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 19]
```diff
@@ -7,289 +7,410 @@
...
+ * @param maskedPan
+ * @since 2.4.0
+ */
+function saveCardAndSubmitToShop(tokenId, maskedPan)
+{
+    var cardSavingRequest = jQuery.ajax({
+        url: ccControllerUrl,
+        data: {
+            action: "saveCard",
+            token_id: tokenId,
```
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Identifier 'token_id' is not in camel case. (camelcase)](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4399691)

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 20]
```diff
@@ -7,289 +7,410 @@
...
+ * @since 2.4.0
+ */
+function getFormData(tokenId = null)
+{
+    var formDataRequest = jQuery.ajax({
+        url: ccControllerUrl,
+        dataType: "json",
+        data: {
+            action: "getSeamlessConfig",
+            cart_id: cartId,
```
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Identifier 'cart_id' is not in camel case. (camelcase)](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4399691)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
           

Coverage increased per file
===========================
+ wirecardpaymentgateway/helper/TemplateHelper.php  100
         

Coverage decreased per file
===========================
- wirecardpaymentgateway/models/Payment.php  -9
- wirecardpaymentgateway/models/PaymentCreditCard.php  -1
         

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/TemplateHelper.php  1
- wirecardpaymentgateway/controllers/front/payment.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/views/js/creditcard.js  -1
         

Clones removed
==============
+ wirecardpaymentgateway/views/js/creditcard.js  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4399691&bid=14911644)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/controllers/front/payment.php  1
- wirecardpaymentgateway/helper/TemplateHelper.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/views/js/creditcard.js  -1
         

Clones removed
==============
+ wirecardpaymentgateway/views/js/creditcard.js  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4399691&bid=14911644)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/controllers/front/payment.php  1
- wirecardpaymentgateway/helper/service/OrderService.php  1
- wirecardpaymentgateway/helper/TemplateHelper.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/views/js/creditcard.js  -1
         

Clones removed
==============
+ wirecardpaymentgateway/views/js/creditcard.js  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4399691&bid=14911644)

**Pull request merged**